### ISSUE TITLE
strongly-recommended rules fails with vue-types

### ISSUE THREAD

**Issue by dwightjack** [vuejs/eslint-plugin-vue#389]
Hi,

thanks for the great effort on this library!

I'm the author of a React's `prop-types`-like library for vue called [vue-types](https://github.com/dwightjack/vue-types).

I've been contacted by a user which reported an [issue](https://github.com/dwightjack/vue-types/issues/29) using `vue-types` with `eslint-plugin-vue` rules.

Basically the library generates a prop object at runtime, but that's not picked up by some rules like `require-default-prop`. If I got it right, since eslint it's a static analyzer, it cannot manage runtime code (even if resolvers like those in eslint-plugin-import seems to somehow be able to do it). 

I'm looking for a way to make this setup work without needing to disable the eslint rule which is pretty important from my point of view.

Any suggestion on where to start?

**Comment by HerringtonDarkholme** [comment]
Libraries other than eslint-plugin-vue rely on props being written as object literal.  For example https://github.com/vuedoc/parser. 

We can teach eslint not to report error when it see non literal value in props, but I wonder how this can be extended to other vue analyzers.

**Comment by dwightjack** [comment]
Hi,

thanks for the reply. manged to take a look into this issue. I'm not an expert on this field, so I'd like to ask whether I'm looking in the right direction.

Other libraries like `eslint-plugin-react` seems to be able to detect library objects like `createClass` and `PureComponent`.

I was wondering if it would be possible to extend eslint (maybe acorn?) by letting it understand `VueTypes` objects and then statically check for properties. 

Let's say I setup .eslintrc.js like this

```js
{
  plugins: [
    'vue-types/eslint-plugin'
  ],
  settings: {
    'vue-types': {
      namespace: 'VueTypes' //the name the package is imported with
    }
  }
}
```

And then somewhere in the parser I will check for `${namespace}.[objectFactory]` and based on that the parser returns a plain prop object.

Not sure I've explain this totally clearly, but I'm really a newbie on this kind of library development.

**Comment by mysticatea** [comment]
Thank you for this issue.

As [ESLint's philosophy](https://eslint.org/docs/about/#philosophy), a rule cannot change the behavior of another rule (though `no-unused-vars` is an only exception). So the solution is to clone a rule into your plugin then use it instead. An ESLint TSC member `not-an-aardvark` has created useful library [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) for this purpose. That `eslint-rule-composer` lets us make a new rule by inheriting an existing rule.

**Comment by dwightjack** [comment]
Thank you for your reply. 

Ok.  I think I could use a namespaced rule like `vue-types/rule-name` . I've seen such pattern in a stylelint plugin and I think it's the cleanest way. 

1) add vue type plugin
2) disable vue-eslint-rule
3) enable vue-types/rule-name rule that inherits and extends vue-eslint-rule

Will look into it. Thanks for the hint!

**Comment by dwightjack** [comment]
Hi,

as an follow up to this issue I've released a plugin to _whitelist_ `vue-types`'s definitions: https://github.com/dwightjack/eslint-plugin-vue-types

Feedback are welcome

**Comment by michalsnik** [comment]
Good job @dwightjack ! Happy you figured it out :) I guess we can close this issue now
